kvstore/encoding.py

```diff
@@ -0,0 +1,118 @@
+from enum import Enum
+import io
+
+class CommandEnum(Enum):
+    GET = 1
+    SET = 2
+    MSET = 3
+    CREATE_TABLE = 4
+    LIST_TABLES = 5
+    REMOVE_TABLE = 6
+
+
+class Command():
+    def encode(self):
+        raise NotImplementedError
+
+
+class BinaryEncoderDecoder:
+    def __init__(self):
+        self.byte_order = 'big'
+        self.string_encoding = 'utf-8'
+
+        self.command_bytes = 2
+        self.command_signed = False
+
+        # TODO: differentiate between str8, str16, and str32
+        self.key_size_bytes = 2
```

Comment:
interesting, does this mean that keys and value strings are a fixed number of bytes?

Reply:
not exactly. the encoding has a "key_size" and then the "key". so the "key_size" field says how many bytes long the "key" is. however, we have to decide in advance how long the "key_size" field will be. i hardcoded a value of 2 bytes for the key_size, which means we could support keys up to 65k bytes long (2^16 possible values specified by key_size). 

i checked out msgpack, and it looks like they do this dynamically, which makes sense. in that encoding, there are 3 types of strings you can have: str8 (length of the string specified by 1 byte), str16 (len specified by 2 bytes), and str32 (len specified by 4 bytes). so with that approach, you need some number of bits (not sure how many msgpack uses) to specify the type of field it is, then the size of the field (either 1, 2, or 4 bytes depending on the type), and then the field itself

with 4 bytes specifying the string size, you can encode strings up to 4 GB, which seems like plenty. but it is interesting to note that encoding like msgpack have a hard limit on string length